packages/databyss-editor/lib/clipboardUtils.tsx

```diff
@@ -0,0 +1,445 @@
+      rangesForBlockBefore.push({
+        offset: r.offset,
+        length: offset - r.offset,
+        marks: r.marks,
+      })
+      rangesForBlockAfter.push({
+        offset: 0,
+        length: r.length + r.offset - offset,
+        marks: r.marks,
+      })
+    }
+    if (r.offset === offset) {
+      rangesForBlockAfter.push({
+        offset: 0,
+        length: r.length,
+        marks: r.marks,
+      })
+    }
+  })
+
+  rangesForBlockBefore = rangesForBlockBefore.filter(r => r.length > 0)
+
+  rangesForBlockAfter = rangesForBlockAfter.filter(r => r.length > 0)
+
+  return {
+    before: {
+      text: {
+        textValue: block.text.textValue.substring(0, offset),
+        ranges: rangesForBlockBefore,
+      },
+      type: block.type,
+    },
+    after: {
+      text: {
+        textValue: block.text.textValue.substring(offset),
+        ranges: rangesForBlockAfter,
+      },
+      type: block.type,
+    },
+  }
+}
+
+// checks is state selection is collapsed
+export const isSelectionCollapsed = (selection: Selection): boolean => {
+  const { anchor, focus } = selection
+  return anchor.index === focus.index && anchor.offset === focus.offset
+}
+
+// return atomic or new id
+const getId = (type: BlockType, id: string): string =>
+  isAtomicInlineType(type) ? id : new ObjectId().toHexString()
+
+// takes blocks array and resets the id's for non atomic types
+export const resetIds = (fragment: Block[]): Block[] =>
+  fragment.map(block => ({ ...block, _id: getId(block.type, block._id) }))
+
+const addBlockData = (frag: Block[]): Block[] =>
+  frag.map(b => ({ ...b, __showNewBlockMenu: false, __isActive: false }))
+
+// always have the anchor come before the focus
+export const sortSelection = (selection: Selection): Selection => {
+  const { anchor, focus } = selection
+  let _anchor = anchor
+  let _focus = focus
+  // always put anchor before focus
+  if (anchor.index > focus.index) {
+    _focus = [_anchor, (_anchor = _focus)][0]
+  } else if (anchor.offset > focus.offset && anchor.index === focus.index) {
+    _focus = [_anchor, (_anchor = _focus)][0]
+  }
+
+  return {
+    anchor: _anchor,
+    focus: _focus,
+  }
+}
+
+// returns fragment in state selection
+export const getCurrentSelection = (state: EditorState): Block[] => {
+  if (isSelectionCollapsed(state.selection)) {
+    return []
+  }
+
+  let frag: Block[] = []
+
+  const { blocks, selection } = state
+
+  const { anchor, focus } = sortSelection(selection)
+
+  const _blocks = cloneDeep(blocks)
+
+  // if selection is within the same block
+  if (anchor.index === focus.index) {
+    const _selectionLength = focus.offset - anchor.offset
+    const _block = blocks[anchor.index]
+    // split block at anchor offset and use `after`
+    const _firstSplit = splitBlockAtOffset({
+      block: _block,
+      offset: anchor.offset,
+    }).after
+
+    // split block at length of selection and get `before`
+    const _secondSplit = splitBlockAtOffset({
```

Comment:
Why do we need 2 splits here?

Reply:
"this is an `|(first split)`example of a selected `|(second split)`fragment"

the selected fragment is `example of a selected`